[SPIRV] Emit DebugFunction/Definition for both wrapper and real functions

SteveUrquhart opened the pull request:
https://github.com/microsoft/DirectXShaderCompiler/pull/6758 unfortunately caused a regression for shaders compiled without -fcgl.  This PR extends the original fix and corrects the regression.  We now emit DebugFunction and DebugFunctionDefinition pairs for both the wrapper and the main functions, and a DebugEntrypoint that points at the wrapper function.  This survives inlining.

s-perron commented:
Lots of tests are failing. Try running the spir-v tests: `ninja check-clang-codegenspirv`.

SteveUrquhart commented:
I'm sorry, I have one test left to fix.

SteveUrquhart commented:
@s-perron, the current failure in CI seems to be coming from an optimizer pass.  I've attached the spvasm if we use fcgl or O0, and we see there are now 2 DebugFunctions, one for the wrapper and one for src.main.  debug_info_manager.cpp:117 in the optimizer asserts.  
    assert(
        fn_id_to_dbg_fn_.find(fn_id) == fn_id_to_dbg_fn_.end() &&
        "Register DebugFunction for a function that already has DebugFunction");
I appreciate your advice on this.
[spirv.debug.opline.end.of.shader.O0.spvasm.txt](https://github.com/user-attachments/files/17423224/spirv.debug.opline.end.of.shader.O0.spvasm.txt)
[spirv.debug.opline.end.of.shader.fcgl.spvasm.txt](https://github.com/user-attachments/files/17423225/spirv.debug.opline.end.of.shader.fcgl.spvasm.txt)

s-perron commented:
@SteveUrquhart The `DebugFunction` instructions do not look correct. Both of them have `%src_main` for their declaration, which is incorrect for two reasons. First they represent different functions, so they should be different. Second, that value should be the result of a `DebugFunctionDeclaration`, and not the result of the `OpFunction`.

```
   %src_main = OpFunction %void None %15
         %10 = OpExtInst %void %1 DebugFunction %8 %5 %6 3 1 %7 %9 FlagIsProtected|FlagIsPrivate 3 %src_main
         %14 = OpExtInst %void %1 DebugFunction %13 %5 %6 3 1 %7 %9 FlagIsProtected|FlagIsPrivate 3 %src_main
```

SteveUrquhart commented:
I've found the NonSemantic DebugScope for the entry function is incorrectly referencing the DebugFunction for the wrapper.  I will enhance the testing to catch this and push again.

SteveUrquhart commented:
I've added testing for NonSemantic DebugScope, and cleaned up the names the debug user sees for the wrapper and the original function.  While debugging src_main, the user will want to see "main" from the HLSL.

s-perron commented:
This looks good to me, but I'm not setup to test it in renderdoc. @Goshido Does this fix #6939?

SteveUrquhart commented:
I've found one more problem with the logic.  The new convenience function emitDebugFunction() was the wrong place to pushCurrentLexicalScope().  I have a fix for this and am adding test logic for it now.

Goshido commented:
I gonna check _PR_ when you are ready.

SteveUrquhart commented:
@Goshido, please check it when you are ready.

Goshido commented:
I checked branch `fix-ns-debugfunctiondef`, commit `e1a4719d7b53ef6fddffa3f18422b6526f65a79b`. _DXC_ version 1.8.2407.93. It's working as expected.
![image](https://github.com/user-attachments/assets/10df1d0c-0835-4f03-818f-65f1523f94b6)

Awesome! 👍

_P.S._ Quick question. What last digit of _DXC_ version means?

```txt
            minor   ???
              |      |
              ▼      ▼
DXC version 1.8.2407.93
            ^     ^
            |     |
          major   |
                  |
             day and month
```
